aquifer/aquifer
# Composer configuration

**Dependent upon #150 and #151 being merged prior as it is build off of those branches.**

Adds mostly out-of-the-box configuration for compatibility with Composer (and also aquifer-git).

Adds dedicated drupal.composer.json for the Drupal build separate from the Aquifer project composer.json.

Performs Composer build by copying `drupal.composer.json`, `drupal.composer.lock`, and `/scripts/composer` into the build directory and runs `composer install` from within the build directory.

This makes `aquifer build` using composer more flexible and able to work with alternative build directories such as those created by `aquifer-git`.

This will break projects that have the build directory prepended to their `vendor-dir` and `installer-paths` settings in `composer.json`. Those settings should be reconfigured to be relative to the project build directory and the file should be renamed `drupal.composer.json`.

You will also need to add the add the following sync items to the `sync` object in your `aquifer.json` file:

```
    "root": {
      "destination": "",
      "method": "symlink",
      "conflict": "overwrite",
      "type": "dir",
      "hook": "postBuild"
    },
    "modules/custom": {
      "destination": "modules/custom",
      "method": "symlink",
      "conflict": "overwrite",
      "type": "dir",
      "hook": "preBuild"
    },
    "themes/custom": {
      "destination": "themes/custom",
      "method": "symlink",
      "conflict": "overwrite",
      "type": "dir",
      "hook": "postBuild"
    },
```

**To test:**
- Create a D8 site by running `aquifer create test -d 8 && cd test`
- Edit the `aquifer.json` file changing the `build` object properties to look like so:
  
  ```
  "build": {
    "method": "composer",
    "directory": "build",
    "makeFile": "composer.json"
  }
  ```
- Run `aquifer build`
- Verify site builds without errors.

**Aquifer Git**
- Note you will likely need to run the 1.0.0 version of aquifer-git to test this successfully. Add the following to `.aquifer/package.json`:
  
  ```
  "dependencies": {
    "aquifer-git": "aquifer/aquifer-git#1.0.0"
  }
  ```
- Run `aquifer extensions-load`
- You will need to have a test depository to deploy to and configure `aquifer-git` accordingly in `aquifer.json`. See [aquifer-git docs](https://github.com/aquifer/aquifer-git/tree/1.0.0).
- Run `aquifer deploy-git`
- Verify command runs without errors and deploys to your repo.
